### DISCUSSION TITLE
test

### DISCUSSION THREAD

**Discussion by daruoxiaoyu** [OpenRealTimeSimulation/SolverCodegen#6] [Q&A]
Hi, thank you very much for your open source contribution.
I'm having some difficulties testing the generated model. Have you tried to convert the generated program into c so that it can be compared with the circuit model in plecs or simulink software.
Thanks again!

**Comment by OpenRealTimeSimulation** [comment]
This discussion has been moved from issues to discussions section of the project.

**Comment by OpenRealTimeSimulation** [comment]
Hello @daruoxiaoyu,

The generated solvers are defined in C++ to allow greater flexibility in defining component model code compared to C.  Moreover, many simulators and FPGA HLS tools support C++ alongside C.  The ORTiS solver codegen tools are not capable of generating solvers within other languages (such as C or Python) at this time.

The C++ defined solvers generated by the ORTiS codegen tools can be called from Simulink and PLECS models.  In Simulink, the solver functions can be called from S-Function blocks.  The C++ solver functions can be used in PLECS by building a DLL around the solver function with a C++ compiler suite and using the built DLL in a DLL block within PLECS.  In both cases, since the solvers are defined as generic function templates in C++ rather than specific functions, the function call --given template parameters for instance and real data type-- will need to be wrapped within another function that is called by Simulink or PLECS.

The component models used in generated solvers from the tool have been tested against Simulink and PLECS before, using the above mentioned approaches or by running the generated solvers in an independent C++ testbench and comparing results after.  A few of the literature (see readme of the project) for the project show such comparison, such as this one (SoftwareX, public access):
https://www.sciencedirect.com/science/article/pii/S2352711021000054

You mentioned you are having difficulties testing a model.  Can you elaborate on what you are trying to do and the issue you are having?

**Reply by CharlinZheng** [reply]
I am interested in the generated function templates for use in Simulink. However, so far I am not very clear about how such function templates can be used in s-function builder. Does the developer have an example that I can use for reference?

**Reply by OpenRealTimeSimulation** [reply]
Hello @ZhouDuu,

The simplest way to use instances of C++ function templates (like the generated solvers) in S-Functions (or in tools like Xilinx Vivado HLS) is to call the parameterized function template from within a regular wrapper function that will be used as the S-function in Simulink.  An example wrapper function would look like this:
```
typedef double real;

void ExampleInverterSystemSolverWrapper
(
real x_out[5],
real& positive_capacitor_voltage_converter,
real& negative_capacitor_voltage_converter,
real& leg_a_inductor_current_converter,
real& leg_b_inductor_current_converter,
real& leg_c_inductor_current_converter,
bool switch_gates_converter[6]
)
{
	//call solver function template for single time step
	//using template parameters 0 for instance and
	//real for real-valued type
	
	ExampleInverterSystem_solver<0,real>
	(
		x_out,
		positive_capacitor_voltage_converter,
		negative_capacitor_voltage_converter,
		leg_a_inductor_current_converter,
		leg_b_inductor_current_converter,
		leg_c_inductor_current_converter,
		switch_gates_converter
	);
}
```
You would call `ExampleInverterSystemSolverWrapper` using the S-function block.  For information on how to use S-Functions in Simulink, refer to this link (MathWorks):
https://www.mathworks.com/help/simulink/s-function-concepts-c.html;jsessionid=55ce011f2713b5f553294208cef0

The solver functions use the LB-LMC solver method (see references in readme file) and are optimized for power electronics real-time simulations with small time steps of 1us or below for FPGA execution using HLS tools.  Depending on your model in the solver function, the solver might not work well at larger time steps.  So you will need to run the Simulink model with the corresponding solver S-Function at same small time step the solver function was configured for (1us or below) for it to work correctly.  Alternatively, you can run the Simulink model at larger time step and have the generated solver function with its smaller time step called multiple times each Simulink time step.  For instance, if the Simulink model's time step is 10us and the generated solver function's time step is 1us, you would call the solver function 10 times each Simulink time step.  This setup can be done simply by having the solver's wrapper call the solver function multiple times within a C++ loop.  In this multi-rate setup, the Simulink and solver function time steps must be integer multiples of one another.

Note that running the generated solver functions in Simulink is primarily used to validate the power electronics model in the solver and test it with various controller/protection logic offline before deploying the solver onto a FPGA with HLS tools to run in real-time with small time steps (like 50ns and below) for HIL testing or power electronics emulation applications.

Hope this information helps you.

BTW, when asking questions on a new topic of your own, please start a new thread in the Discussions page.  In this way, your question and its answers are more visible to us and others and is not buried within another person's thread.

**Reply by CharlinZheng** [reply]
Thank you very much for your help, I have successfully run the LB-LMC based model on Simulink's S-function builder. As a side note, it is easier to use LB-LMC solver, using S-function-builder. There is no need to configure S-function.
By the way, I will put the above issues in a new discussion for more people to see clearly.

**Reply by xbh-sdju** [reply]
Hello, can you take a look at how your simulink -builder is used, I have found some information on the Internet without success.

**Comment by xbh-sdju** [comment]
I've generated h files with CLI tools, and c++ files written through the instructables, but no CSV files are output, which is bad. What is the solution ?

**Reply by OpenRealTimeSimulation** [reply]
Hello @xbh-sdju,

To better help you, please start a new discussion describing the issue you are having.  In the new discussion, please describe in full detail what you are trying to do and the issue you are having.  If possible, provide copies of your code there to clarify what you are trying to accomplish.

Thank you.

**Reply by xbh-sdju** [reply]
[rlc2.zip](https://github.com/OpenRealTimeSimulation/SolverCodegen/files/8548954/rlc2.zip)
 
When I write an offline C++ emulation test file in Microsoft Visual C++ 2010 Express, it shows that my RLC_Circuit.hpp is not found, but I have it loaded. This is really bad, RLC_Circuit.hpp is generated from the CLI file and the example is the RLC example from the operation manual. Hope to get your help. Thank you very much, the attachment is my project file.

**Comment by OpenRealTimeSimulation** [comment]
Discussion closed and locked due to inactivity.
